helix-core/src/main/java/org/apache/helix/task/ThreadCountBasedTaskAssignmentCalculator.java

```diff
@@ -65,10 +66,26 @@ public ThreadCountBasedTaskAssignmentCalculator(TaskAssigner taskAssigner,
       Map<String, IdealState> idealStateMap) {
     Map<String, TaskConfig> taskMap = jobCfg.getTaskConfigMap();
     Map<String, Integer> taskIdMap = jobCtx.getTaskIdPartitionMap();
+    // Check if a gap exists in the context due to previouslys
+    // removed tasks. If yes, the missing pIDs should be considered
+    // for newly added tasks
+    Set<Integer> existingPartitions = jobCtx.getPartitionSet();
+    Set<Integer> missingPartitions = new HashSet<>();
+    if (existingPartitions.size() != 0) {
+      for (int pId = 0; pId < Collections.max(existingPartitions); pId++) {
+        if (!existingPartitions.contains(pId)) {
+          missingPartitions.add(pId);
+        }
+      }
+    }
     for (TaskConfig taskCfg : taskMap.values()) {
       String taskId = taskCfg.getId();
-      int nextPartition = jobCtx.getPartitionSet().size();
       if (!taskIdMap.containsKey(taskId)) {
+        int nextPartition = jobCtx.getPartitionSet().size();
+        if (missingPartitions.size()!=0) {
+          nextPartition = missingPartitions.iterator().next();
```

Comment:
Is it better to instead start with "the largest partitionID in jobCtx", and simply +1 every iteration? The fill-the-gaps approach seems unnecessarily complicated, if I'm not missing anything.

Reply:
So the issue here is:
Let's assume we have these tasks [0,1,2,3,4]
Now the user deletes task 3 and we will have [0,1,2,4]
Now if the new task is being added, the size is 4 and the new pId will be 4 instead of 3. Hence the task would be be on context as we have a collision.
That is why we need to find the missing tasks.

Reply:
I have concerns for you are using missing partitions. Let's say user removed 3 as you mentioned, the job is not fully cancelled and still in running stage. If they add another task and you reuse missing partition 3 as the job id. Will that maps to existing non-cancelled task and make the system in a wired state?

Reply:
That would not be the case. The design of the delete is to remove the task from context after it has been dropped completely and the current state is gone. It means after the cancellation we will remove context. In this case, it is safe to add a new tass with ID=3.